### DISCUSSION TITLE
Partial match for dynamic routes

### DISCUSSION THREAD

**Discussion by WingmanImd** [vercel/next.js#15690] [Ideas]
In a project I have the folowing page structure:
/
/[lang]
/[lang]/index.js 
/[...all].js

This works fine until I had to also accept urls like: "/cms-page.html"(where "cms-page" is a variable). Obviously I have a match on /[lang]/index.js and I can, based on the query to check if it's and cms page or the normal page with a language selected.
I would love to have partial matching for routes so I can do stuff like:
/[lang]/[cmsPageName].html.js - matches any url that is 2 level deep and ends with ".html"
/faq-[faq-name].js - matches any url that starts with "faq-"
/faq-[faq-name]-page.js
/faq-[faq-name][faq-id].js

I think this will make the routing system a lot more powerful and would make a lot of developers's lives a lot easier.
At the moment the only way to do this is to create a lot of [...all] pages and based on the url return the page needed. The downside is that that page will contain all the possible pages and the bundle for those routes will be quite big(ex: for 5+ possible return values)

**Comment by svobik7** [comment]
@WingmanImd I like your idea to have partial matches. In my case I would use something like `/@[username]` for user detail.

The workaround as you wrote can be enhaced by `next/dynamic` so that only the right page will be included.

**Reply by WingmanImd** [reply]
Thanks for the reply!
But next/dynamic doesn't work on SSR as far as I know, at least it didn't work for me.
I ended up using a custom server, a custom _app and some minor abuse of [...all] pages as my project also required to have totally different pages for desktop and mobile for the same url and in the same app with SSR being a requirement for both.

**Reply by svobik7** [reply]
According to docs https://nextjs.org/docs/advanced-features/dynamic-import the dynamic imports work well during SSR. I am just in the middle of implementing `[...slug]` page in my new project so I can let you know if it works.

**Comment by sbettwieser** [comment]
I really hope some dev sees this post and makes it a reality. I really want to do things like `/pages/obj-[objectName].xml.js`.

**Comment by neojpk** [comment]
please nextjs team this could be paired up with socket-io dynamic namespaces like room-[id] for an awesome lib interop

**Comment by vinit03** [comment]
I had the same requirement. Next JS does not provide this functionality but you could achieve the same using next js rewrites:

Basically I had 3 types of pages. All Resources Page, All Article Page and All Video Page

My folder structure:
```
├── [resourceCategory]
│   ├── article
│   │   └── [[...page]]
│   │       └── page.tsx
│   ├── page.tsx
│   └── videos
│       └── [[...page]]
│           └── page.tsx
```

Based on the folder structure my valid next URLs were:

All resource URLs:
/resources
/resources-p
/<resource>

All article URLs:
/resources/articles
/resources/articles/2 (paginated)
/resources-p/articles
/resources-p/articles/3 (paginated)
/<resources>/articles
/<resources>/articles/2 (paginated)

All video URLs:
/resources/videos
/resources/videos/2 (paginated)
/resources-p/videos
/resources-p/videos/3 (paginated)
/<resources>/videos
/<resources>/videos/2 (paginated)


My requirement:
All resource actual url should be:
/resources
/resources-<filter_name>/<filter_slug>

All article actual url should be:
/resources/articles
/resources-<filter_name>/<filter_slug>/articles

All video actual url should be:
/resources/videos
/resources-<filter_name>/<filter_slug>/videos


This is my rewrite in next.config.js usign which I was able to achieve what I wanted without update my dir structure:
```
async rewrites() {
    return [
      {
        source: "/resources-:filterType/:filterSlug",
        destination: "/resources-:filterType?filterSlug=:filterSlug",
      },
      {
        source: "/resources-:filterType/:filterSlug/articles",
        destination: "/resources-:filterType/articles?filterSlug=:filterSlug",
      },
      {
        source: "/resources-:filterType/:filterSlug/articles/:page",
        destination: "/resources-:filterType/articles/:page?filterSlug=:filterSlug",
      },
      {
        source: "/resources-:filterType/:filterSlug/videos",
        destination: "/resources-:filterType/videos?filterSlug=:filterSlug",
      },
      {
        source: "/resources-:filterType/:filterSlug/videos/:page",
        destination: "/resources-:filterType/videos/:page?filterSlug=:filterSlug",
      },
    ];
  },
  ```
